sky/provision/gcp/instance.py

```diff
@@ -169,17 +173,65 @@ def terminate_instances(
     # time (same as what we did in ray's node_provider).
 
 
+def open_ports(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, cluster_name_on_cloud
+    zone = provider_config['availability_zone']
+    project_id = provider_config['project_id']
+    firewall_rule_name = provider_config['firewall_rule']
+
+    label_filters = {TAG_RAY_CLUSTER_NAME: cluster_name_on_cloud}
+    handlers: List[Type[instance_utils.GCPInstance]] = [
+        instance_utils.GCPComputeInstance
+    ]
+    handler_to_instances = _filter_instances(handlers, project_id, zone,
+                                             label_filters, lambda _: None)
+    operations = collections.defaultdict(list)
+    for handler, instances in handler_to_instances.items():
+        if not instances:
+            logger.warning(f'No instance found for cluster '
+                           f'{cluster_name_on_cloud}.')
+            continue
+        # If multiple instances are found, they are in the same cluster,
+        # i.e. the same VPC. So we can just pick one.
+        try:
+            vpc_name = handler.get_vpc_name(project_id, zone, instances[0])
+        except gcp.http_error_exception() as e:
+            if _INSTANCE_RESOURCE_NOT_FOUND_PATTERN.search(e.reason) is None:
+                logger.warning(f'Failed to get VPC name for instance '
+                               f'{instances[0]}. Skip opening ports for it.')
+            else:
+                logger.warning(f'Instance {instances[0]} does not exist. '
+                               f'Skip opening ports for it.')
+        else:
+            op = handler.create_or_update_firewall_rule(
+                firewall_rule_name,
+                project_id,
+                cluster_name_on_cloud,
+                ports,
+                vpc_name,
+            )
```

Comment:
Seems the vpc_name is only used by the `create_or_update_firewall_rule`. Can we combine the two functions, i.e. having the latter take the `instance`.
Design Principle: hide as many details as possible for an API.